dfu: multi image package library (#7493)
nrfconnect/sdk-nrf: doc/nrf/libraries/dfu/dfu_multi_image.rst

.. _lib_dfu_multi_image:

DFU Multi Image
###############

.. contents::
   :local:
   :depth: 2

The DFU Multi Image library provides an API for writing a DFU Multi Image package that is being downloaded in chunks of an arbitrary size.

> The DFU Multi Image package is a simple archive file consisting of a CBOR-based header describing contents of the package, followed by a number of updates images, such as firmware images for different MCU cores.

Reviewer: ```suggestion
The DFU multi-image package is a simple archive file that consists of a CBOR-based header that describes contents of the package, followed by a number of updates images, such as firmware images for different MCU cores.
```

> Images included in a DFU Multi Image package are identified by user-assigned numeric identifiers.

Reviewer: ```suggestion
Images included in a DFU multi-image package are identified by numeric identifiers assigned by the user.
```

> The library provides a means for a user to register custom functions for writing package data that belongs to an image with a given identifier.

Reviewer: ```suggestion
The library provides a way for the user to register custom functions for writing package data.
The data is assigned to an image with a given identifier.
```
Just trying to break the sentence into smaller chunks...

> That makes the library a general-purpose solution as it makes no assumptions with respect to formats of images included in a written package.

Reviewer: ```suggestion
Given that the library makes no assumptions with respect to formats of images included in a written package, it makes it a general-purpose solution for multi-image builds, for example for the :ref:`nRF5340 DK <ug_nrf5340_multi_image>`.
```

Author: I rephrased a bit to convey better what I meant. I would like to not make the impression that the library is only for multi-image builds (meaning multi-image firmwares).

Configuration
*************

To enable the DFU Multi Image library, set the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE` Kconfig option.

To configure the maximum number of images that the DFU Multi Image library is able to process, use the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT` Kconfig option.

To enable building the DFU Multi Image package that contains common update images, such as the application core firmware, the network core firmware, or MCUboot images, set the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE_PACKAGE_BUILD` Kconfig option.

> 

Reviewer: Dependencies are required by the library template, so maybe let's add this section here with a link to zcbor?
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/library_template_README.html#dependencies

Author: Ok, added.

.. _lib_dfu_multi_image:

DFU multi-image
###############

.. contents::
   :local:
   :depth: 2

The DFU multi-image library provides an API for writing a DFU multi-image package that is being downloaded in chunks of an arbitrary size.
The DFU multi-image package is a simple archive file that consists of a CBOR-based header that describes contents of the package, followed by a number of updates images, such as firmware images for different MCU cores.

> Images included in a DFU multi-image package are identified by numeric identifiers assigned by the user.
> The library provides a way for the user to register custom functions for writing a single image with a given identifier.

Reviewer 2: ```suggestion

The library provides a way for the user to register custom functions for writing a single image and assigning it with a numeric identifier.
Images included in a DFU multi-image package are identified by this numeric identifier.

```
Note the empty lines before and after. It'll improve readability to separate the text from one long paragraph.

> Given that the library makes no assumptions with respect to formats of images included in a written package, it makes it a general-purpose solution for device firmware upgrades.

Reviewer 2: ```suggestion
Because the library makes no assumptions about the formats of images included in a written package, it serves as a general-purpose solution for device firmware upgrades.
```

For example, it can be used to upgrade the :ref:`nRF5340 DK <ug_nrf5340_multi_image>` firmware.

Configuration
*************

To enable the DFU multi-image library, set the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE` Kconfig option.

To configure the maximum number of images that the DFU multi-image library is able to process, use the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT` Kconfig option.

> To enable building the DFU multi-image package that contains common update images, such as the application core firmware, the network core firmware, or MCUboot images, set the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE_PACKAGE_BUILD` Kconfig option.

Reviewer 2: ```suggestion
To enable building a DFU multi-image package that contains common update images, such as the application core firmware, the network core firmware, or MCUboot images, set the :kconfig:option:`CONFIG_DFU_MULTI_IMAGE_PACKAGE_BUILD` Kconfig option.
```
I'm not sure what the word "common" is doing here? Does it means "shared"?

Author: It means "commonly used".  Typically people want to include the application core/network core firmware when they perform the DFU. Should I use "commonly used" or you have a better suggestion?

Reviewer 2: Yes, "commonly used" is much clearer.
